### PR TITLE
Fix Traces functionality

### PR DESCRIPTION
**Pull request by jtkrogel**
## Proposed changes

Traces became broken due to changes in hdf_archive.  These were not caught by the current 1 mpi 1 omp-thread deterministic traces tests.  

This PR adds a 4 mpi 4 omp-thread deterministic test for traces and updates hdf_archive to allow Traces to function again.  

## What type(s) of changes does this code introduce?

- Bugfix

### Does this introduce a breaking change?

- No

## What systems has this change been tested on?

Laptop

## Checklist

- Yes. This PR is up to date with current the current state of 'develop'

### CODE REVIEW

File: src/io/hdf/hdf_archive.cpp [hunk 1]
```diff
@@ -142,8 +142,8 @@ bool hdf_archive::create(const std::filesystem::path& fname, unsigned flags)
   possible_filename_ = fname;
   if (Mode[NOIO])
     return true;
-  if (!(Mode[IS_PARALLEL] || Mode[IS_MASTER]))
-    throw std::runtime_error("Only create file in parallel or by master but not every rank!");
```
Comment (ye-luo) [inline src/io/hdf/hdf_archive.cpp:146]: Why are these lines offensive? Please explain your use case.

Comment (ye-luo) [reply src/io/hdf/hdf_archive.cpp:146]: I feel the caller side was using hdf_archive incorrectly.

Comment (jtkrogel) [reply src/io/hdf/hdf_archive.cpp:146]: This is what blocks Traces from functioning in a parallel run.  Traces creates one H5 file per MPI process.  This does not create significant runtime overhead for QMCPACK vs having traces off (measured at about a 5% performance impact on large scale Titan runs).

Comment (jtkrogel) [reply src/io/hdf/hdf_archive.cpp:146]: See Sec. IV A in https://doi.org/10.1109/CSE.2013.24

Comment (ye-luo) [reply src/io/hdf/hdf_archive.cpp:146]: hdf_archive should support one file file per rank when it is not aware of MPI. So the issue is incorrect open mode of the file not the hdf_archive itself. Please provide a link where open() is called from the trace code.

Comment (jtkrogel) [reply src/io/hdf/hdf_archive.cpp]: This works.  I've restored the guards in hdf_archive now and the update for the fix is now confined to TraceManager.h

**Comment by ye-luo** [comment]
Test this please

**Review by ye-luo — APPROVED** [review]

**Pull request merged**